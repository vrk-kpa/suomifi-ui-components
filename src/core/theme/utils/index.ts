export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
