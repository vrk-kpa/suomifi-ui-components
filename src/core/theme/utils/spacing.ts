import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

export type GlobalMarginProps = { globalMargins: GlobalMargins };

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export type GlobalMargins = {
  all?: MarginProps;
  alert?: MarginProps;
  actionMenu?: MarginProps;
  block?: MarginProps;
  button?: MarginProps;
  breadcrumb?: MarginProps;
  checkbox?: MarginProps;
  checkboxGroup?: MarginProps;
  chip?: MarginProps;
  dateInput?: MarginProps;
  dropdown?: MarginProps;
  expander?: MarginProps;
  expanderGroup?: MarginProps;
  externalLink?: MarginProps;
  fileInput?: MarginProps;
  heading?: MarginProps;
  hintText?: MarginProps;
  inlineAlert?: MarginProps;
  label?: MarginProps;
  languageMenu?: MarginProps;
  link?: MarginProps;
  linkList?: MarginProps;
  loadingSpinner?: MarginProps;
  multiSelect?: MarginProps;
  notification?: MarginProps;
  pagination?: MarginProps;
  paragraph?: MarginProps;
  radioButton?: MarginProps;
  radioButtonGroup?: MarginProps;
  routerLink?: MarginProps;
  searchInput?: MarginProps;
  serviceNavigation?: MarginProps;
  sideNavigation?: MarginProps;
  singleSelect?: MarginProps;
  staticChip?: MarginProps;
  statusText?: MarginProps;
  text?: MarginProps;
  textarea?: MarginProps;
  textInput?: MarginProps;
  timeInput?: MarginProps;
  toast?: MarginProps;
  toggleInput?: MarginProps;
  toggleButton?: MarginProps;
  tooltip?: MarginProps;
  wizardNavigation?: MarginProps;
};

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (
  spacing?: SpacingProps,
  important?: boolean,
): string => {
  if (!spacing) return '';

  const cssStyles = Object.entries(spacing)
    .map(([key, value]) =>
      getCSSSpacing(
        defaultSuomifiTheme,
        key as keyof SpacingProps,
        value,
        important,
      ),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
  important: boolean = false,
) => {
  const amount = spaceVal(theme)(value);
  const importantValue = important ? '!important' : '';
  switch (key) {
    case 'mx':
      return `margin-right: ${amount} ${importantValue}; margin-left: ${amount} ${importantValue};`;
    case 'my':
      return `margin-top: ${amount} ${importantValue}; margin-bottom: ${amount} ${importantValue};`;
    case 'px':
      return `padding-right: ${amount} ${importantValue}; padding-left: ${amount} ${importantValue};`;
    case 'py':
      return `padding-top: ${amount} ${importantValue}; padding-bottom: ${amount} ${importantValue};`;
    default:
      return `${cssSelector[key]}: ${amount} ${importantValue};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
