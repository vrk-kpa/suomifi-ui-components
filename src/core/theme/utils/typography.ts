import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
