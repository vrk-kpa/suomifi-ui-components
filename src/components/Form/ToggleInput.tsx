import React, {
  Component,
  ReactElement,
  ComponentClass,
  FunctionComponent,
} from 'react';
import classnames from 'classnames';
import { HtmlInput, HtmlLabel } from '../../reset';
import { ToggleProps, ToggleState } from './Toggle';
import styled from 'styled-components';
import { disabledCursor } from '../utils/css';

const baseClassName = 'fi-toggle';
const toggleDisabledClassName = `${baseClassName}--disabled`;

const componentOrElementWithProps = (
  component: ReactElement<any> | FunctionComponent<any> | ComponentClass<any>,
  props: object,
) => {
  if (!!component) {
    if (React.isValidElement(component)) {
      return React.cloneElement(component, props); // element
    }
    return React.createElement(
      component as FunctionComponent<any> | ComponentClass<any>,
      props,
    ); // component
  }
  return;
};

const StyledHtmlLabel = styled(HtmlLabel)`
  &.${toggleDisabledClassName} {
    ${disabledCursor}
  }
  cursor: pointer;
`;

export class ToggleInput extends Component<ToggleProps> {
  state: ToggleState = {
    toggleState: !!this.props.checked || !!this.props.defaultChecked || false,
  };

  componentWillReceiveProps(nextProps: ToggleProps) {
    const { checked } = nextProps;
    if (!!checked) {
      this.setState({ toggleState: !!checked });
    }
  }

  handleClick = () => {
    const { checked, onClick } = this.props;
    const { toggleState } = this.state;
    if (checked === undefined) {
      this.setState({ toggleState: !toggleState });
    }
    if (!!onClick) {
      onClick({ toggleState: !toggleState });
    }
  };

  render() {
    const {
      id,
      className,
      disabled = false,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      children,
      toggleInputProps,
      toggleInputComponent,
      checked: dissMissChecked,
      defaultChecked: dissMissDefaultChecked,
      onClick: dissMissOnClick,
      ...passProps
    } = this.props;
    const { toggleState } = this.state;
    const toggleClassName = `${baseClassName}--with-input`;
    const toggleInputClassName = `${baseClassName}_input`;
    const toggleLabelClassName = `${baseClassName}_label`;

    const newToggleInputProps = {
      disabled,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      checked: !!toggleState,
      className: toggleInputClassName,
      onChange: this.handleClick,
      ...toggleInputProps,
      id,
    };

    return (
      <StyledHtmlLabel
        htmlFor={id}
        className={classnames(
          toggleClassName,
          className,
          baseClassName,
          {
            [toggleDisabledClassName]: !!disabled,
          },
          toggleLabelClassName,
        )}
        {...passProps}
      >
        {!!toggleInputComponent ? (
          componentOrElementWithProps(toggleInputComponent, newToggleInputProps)
        ) : (
          <HtmlInput {...newToggleInputProps} type="checkbox" />
        )}
        {children}
      </StyledHtmlLabel>
    );
  }
}
