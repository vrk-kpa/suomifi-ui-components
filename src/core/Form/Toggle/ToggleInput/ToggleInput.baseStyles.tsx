import { css } from 'styled-components';
import { SuomifiTheme } from '../../../theme';
import { element, font } from '../../../theme/reset';
import {
  toggleBaseStyles,
  iconWidth,
  iconHeight,
  focusOverrides,
} from '../ToggleBase/Toggle.baseStyles';
import { MarginProps, buildSpacingCSS } from '../../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${toggleBaseStyles(theme)}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)};
  &.fi-toggle--input {
    &:focus-within {
      outline: 0;
      & .fi-toggle_icon-container {
        &:after {
          ${theme.focuses.absoluteFocus}
          ${focusOverrides}
            ${theme.focuses.highContrastFocus} /* For high contrast mode */
        }
      }
    }
    & .fi-text {
      margin: 0;
    }
  }

  & .fi-toggle_input-element {
    ${element(theme)}
    ${font(theme)('bodyText')}
    position: absolute;
    width: ${iconWidth};
    height: ${iconHeight};
    opacity: 0;
    z-index: -9999;
    background-color: ${theme.colors.whiteBase};
  }
`;
