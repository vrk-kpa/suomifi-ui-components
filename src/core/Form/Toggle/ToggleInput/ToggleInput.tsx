import React, { Component, forwardRef } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { AutoId } from '../../../utils/AutoId/AutoId';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../../theme/utils/spacing';
import { getConditionalAriaProp } from '../../../../utils/aria';
import { Text } from '../../../Text/Text';
import {
  HtmlLabel,
  HtmlSpan,
  HtmlInputProps,
  HtmlInput,
} from '../../../../reset';
import { ToggleIcon } from '../ToggleBase/ToggleIcon';
import { ToggleBaseProps, baseClassName } from '../ToggleBase/ToggleBase';
import { baseStyles } from './ToggleInput.baseStyles';
import { filterDuplicateKeys } from '../../../../utils/common/common';

const toggleClassNames = {
  disabled: `${baseClassName}--disabled`,
  input: `${baseClassName}--input`,
  inputElement: `${baseClassName}_input-element`,
  label: `${baseClassName}_label`,
  iconContainer: `${baseClassName}_icon-container`,
};

interface ToggleState {
  toggleState: boolean;
}

export interface ToggleInputProps
  extends ToggleBaseProps,
    MarginProps,
    Omit<HtmlInputProps, 'onChange' | 'type'> {
  /** HTML name attribute for the input */
  name?: string;
  /** Callback fired on input value change */
  onChange?: (checked: boolean) => void;
  /** Ref object is forwarded to the underlying input element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLInputElement>;
}

class BaseToggleInput extends Component<ToggleInputProps> {
  state: ToggleState = {
    toggleState: !!this.props.checked || !!this.props.defaultChecked,
  };

  static getDerivedStateFromProps(
    nextProps: ToggleBaseProps,
    prevState: ToggleState,
  ) {
    const { checked } = nextProps;
    if (checked !== undefined && checked !== prevState.toggleState) {
      return { toggleState: checked };
    }
    return null;
  }

  handleChange = () => {
    const { checked, onChange } = this.props;
    const { toggleState } = this.state;
    if (checked === undefined) {
      this.setState({ toggleState: !toggleState });
    }
    if (!!onChange) {
      onChange(!toggleState);
    }
  };

  render() {
    const {
      children,
      disabled = false,
      onChange,
      id,
      name,
      className,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      checked,
      defaultChecked: dissMissDefaultChecked,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    const { toggleState } = this.state;

    return (
      <HtmlSpan
        className={classnames(
          className,
          baseClassName,
          toggleClassNames.input,
          {
            [toggleClassNames.disabled]: !!disabled,
          },
        )}
        style={style}
      >
        <HtmlLabel className={toggleClassNames.label} htmlFor={id}>
          <HtmlInput
            {...passProps}
            checked={!!toggleState}
            id={id}
            name={name}
            disabled={disabled}
            className={toggleClassNames.inputElement}
            onChange={this.handleChange}
            type="checkbox"
            {...getConditionalAriaProp('aria-label', [ariaLabel])}
            {...getConditionalAriaProp('aria-labelledby', [ariaLabelledBy])}
          />
          <HtmlSpan className={toggleClassNames.iconContainer}>
            <ToggleIcon disabled={disabled} checked={toggleState} />
          </HtmlSpan>
          <Text color={!!disabled ? 'depthBase' : 'blackBase'}>{children}</Text>
        </HtmlLabel>
      </HtmlSpan>
    );
  }
}

const StyledToggleInput = styled(
  (props: ToggleInputProps & SuomifiThemeProp & GlobalMarginProps) => {
    const { theme, globalMargins, ...passProps } = props;
    return <BaseToggleInput {...passProps} />;
  },
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.toggleButton,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const ToggleInput = forwardRef(
  (props: ToggleInputProps, ref: React.RefObject<HTMLInputElement>) => {
    const { id: propId, ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <AutoId id={propId}>
                {(id) => (
                  <StyledToggleInput
                    theme={suomifiTheme}
                    id={id}
                    globalMargins={margins}
                    forwardedRef={ref}
                    {...passProps}
                  />
                )}
              </AutoId>
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

ToggleInput.displayName = 'ToggleInput';
export { ToggleInput };
