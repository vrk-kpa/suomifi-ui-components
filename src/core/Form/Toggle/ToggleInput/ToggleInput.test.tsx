import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { axeTest } from '../../../../utils/test';

import { ToggleInput } from './ToggleInput';

describe('Basic ToggleInput', () => {
  it('should match snapshot', () => {
    const TestToggle = (children: string) => (
      <ToggleInput>{children}</ToggleInput>
    );
    const { container } = render(TestToggle('TestToggle'));
    expect(container.firstChild).toMatchSnapshot();
  });

  it('calling render with the same component on the same container does not remount', () => {
    const TestToggle = (children: string) => (
      <ToggleInput>{children}</ToggleInput>
    );
    const { getByText, rerender } = render(TestToggle('TestToggle'));
    expect(getByText('TestToggle')).toBeTruthy();

    // re-render the same component with different props
    rerender(TestToggle('Test Two'));
    expect(getByText('Test Two')).toBeTruthy();
  });

  it('onClick should activate toggle and deactivate when clicked again', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole, container } = render(
      <ToggleInput onChange={mockClickHandler}>Test two</ToggleInput>,
    );
    const toggle = getByRole('checkbox');
    fireEvent.click(toggle);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    expect(toggle).toBeChecked();
    const svgClassList = container.querySelector('svg')?.classList;
    expect(svgClassList).toContain('fi-toggle_icon');
    expect(svgClassList).toContain('fi-toggle_icon--checked');
    fireEvent.click(toggle);
    expect(svgClassList).not.toContain('fi-toggle_icon--checked');
    expect(toggle).not.toBeChecked();
  });

  it('onClick should not change toggle state when checked is controlled', async () => {
    const mockClickHandler = jest.fn();
    const { getByRole, container } = render(
      <ToggleInput checked={false} onChange={mockClickHandler}>
        Controlled checked state
      </ToggleInput>,
    );
    const toggle = getByRole('checkbox');
    fireEvent.click(toggle);
    expect(mockClickHandler).toHaveBeenCalledTimes(1);
    expect(toggle).not.toBeChecked();
    const svgClassList = container.querySelector('svg')?.classList;
    expect(svgClassList).toContain('fi-toggle_icon');
    expect(svgClassList).not.toContain('fi-toggle_icon--checked');
  });

  test(
    'should not have basic accessibility issues',
    axeTest(<ToggleInput>Toggle</ToggleInput>),
  );
});

describe('name', () => {
  const { getByRole } = render(<ToggleInput name="testToggle" />);
  const namedToggle = getByRole('checkbox') as HTMLInputElement;
  it('has the given name attribute', () => {
    expect(namedToggle.name).toBe('testToggle');
  });
});

describe('margin', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(<ToggleInput margin="xs" />);
    expect(container.firstChild).toHaveStyle('margin: 10px');
  });

  it('should have margin prop overwritten by style prop', () => {
    const { container } = render(
      <ToggleInput margin="xs" style={{ margin: 2 }} />,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});
