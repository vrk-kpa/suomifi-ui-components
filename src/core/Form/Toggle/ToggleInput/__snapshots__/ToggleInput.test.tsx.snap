// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic ToggleInput should match snapshot 1`] = `
.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c3::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c6 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  color: hsl(0,0%,13%);
}

.c6.fi-text--bold {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 600;
}

.c6.fi-text--lead {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 22px;
  line-height: 1.5;
  font-weight: 400;
}

.c6.fi-text--small-screen {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c6.fi-text--small-screen.fi-text--bold {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c6.fi-text--small-screen.fi-text--lead {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 20px;
  line-height: 1.5;
  font-weight: 400;
}

.c4 {
  vertical-align: baseline;
}

.c4.fi-static-icon {
  display: inline-block;
}

.c4.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c4.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c5 {
  width: 40px;
  height: 24px;
  vertical-align: bottom;
  overflow: visible;
}

.c5 .fi-icon-toggle-base {
  stroke: hsl(201,7%,58%);
  fill: hsl(0,0%,100%);
}

.c5 .fi-icon-toggle-knob {
  fill: hsl(201,7%,46%);
  -webkit-transform: translateX(-17px);
  -ms-transform: translateX(-17px);
  transform: translateX(-17px);
}

.c5 .fi-icon-toggle-checkmark {
  fill: none;
}

.c5.fi-toggle_icon--checked .fi-icon-toggle-base {
  stroke: hsl(201,7%,58%);
}

.c5.fi-toggle_icon--checked .fi-icon-toggle-knob {
  fill: hsl(166,90%,34%);
  -webkit-transform: translateX(0px);
  -ms-transform: translateX(0px);
  transform: translateX(0px);
}

.c5.fi-toggle_icon--checked .fi-icon-toggle-checkmark {
  fill: hsl(0,0%,100%);
}

.c5.fi-toggle_icon--disabled .fi-icon-toggle-base {
  stroke: hsl(202,7%,80%);
  fill: hsl(202,7%,97%);
}

.c5.fi-toggle_icon--disabled .fi-icon-toggle-knob {
  fill: hsl(202,7%,80%);
}

.c5.fi-toggle_icon--disabled .fi-icon-toggle-checkmark {
  fill: none;
}

.c5.fi-toggle_icon--checked.fi-toggle_icon--disabled .fi-icon-toggle-checkmark {
  fill: hsl(0,0%,100%);
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  padding-left: 50px;
  position: relative;
  cursor: pointer;
}

.c1.fi-toggle {
  display: inline-block;
}

.c1.fi-toggle--disabled {
  cursor: not-allowed;
}

.c1 .fi-toggle_icon-container {
  position: absolute;
  margin-right: 10px;
  left: 0;
  top: 0.1em;
}

.c1.fi-toggle--input:focus-within {
  outline: 0;
}

.c1.fi-toggle--input:focus-within .fi-toggle_icon-container:after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 14px;
  right: -4px;
  left: -4px;
  outline: 3px solid transparent;
}

.c1.fi-toggle--input .fi-text {
  margin: 0;
}

.c1 .fi-toggle_input-element {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
  position: absolute;
  width: 40px;
  height: 24px;
  opacity: 0;
  z-index: -9999;
  background-color: hsl(0,0%,100%);
}

<span
  class="c0 c1 fi-toggle fi-toggle--input"
>
  <label
    class="c2 fi-toggle_label"
    for="3"
  >
    <input
      class="c3 fi-toggle_input-element"
      id="3"
      type="checkbox"
    />
    <span
      class="c0 fi-toggle_icon-container"
    >
      <svg
        aria-hidden="true"
        class="fi-icon c4 fi-toggle_icon c5"
        focusable="false"
        height="1em"
        viewBox="0 0 37 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <rect
            class="fi-icon-toggle-base"
            height="13"
            rx="6.5"
            width="36"
            x="0.5"
            y="3.5"
          />
          <circle
            class="fi-icon-toggle-knob"
            cx="27"
            cy="10"
            r="10"
          />
          <path
            class="fi-icon-toggle-checkmark"
            d="M31.724 7.692 26 13.71a.91.91 0 0 1-1.333 0l-2.391-2.512a1.03 1.03 0 0 1 0-1.403.911.911 0 0 1 1.333 0l1.724 1.813 5.057-5.316a.91.91 0 0 1 1.334 0 1.026 1.026 0 0 1 0 1.401"
          />
        </g>
      </svg>
    </span>
    <span
      class="c0 fi-text c6 fi-text--body"
    >
      TestToggle
    </span>
  </label>
</span>
`;
